zebra-utils/zcash-rpc-block-template-to-proposal

```diff
@@ -0,0 +1,201 @@
+        echo
+    fi
+
+    if [ "$PROPOSAL_HEIGHT" -ne "$TEMPLATE_HEIGHT" ]; then
+        echo "WARNING: proposing block templates at different heights:"
+        echo "$TEMPLATE_NODE (RPC port $TEMPLATE_RPC_PORT) template is on: $TEMPLATE_HEIGHT"
+        echo "RPC port $PORT proposal is on: $PROPOSAL_HEIGHT"
+        echo
+    fi
+done
+
+echo
+
+TEMPLATE_NODE_TEMPLATE_RESPONSE="$ZCASH_RPC_TMP_DIR/template-getblocktemplate-template.json"
+
+echo "getblocktemplate template request ($TEMPLATE_NODE port $TEMPLATE_RPC_PORT):"
+echo "getblocktemplate $TEMPLATE_ARG_FULL"
+echo
+
+echo "Querying $TEMPLATE_NODE $TEMPLATE_NET chain at height >=$TEMPLATE_HEIGHT..."
+time $ZCASH_CLI -rpcport="$TEMPLATE_RPC_PORT" getblocktemplate "$TEMPLATE_ARG_FULL" > "$TEMPLATE_NODE_TEMPLATE_RESPONSE"
+
+echo "Block template data is in $TEMPLATE_NODE_TEMPLATE_RESPONSE"
+#cat "$TEMPLATE_NODE_TEMPLATE_RESPONSE"
+
+echo
+
+PROPOSAL_DATA="$ZCASH_RPC_TMP_DIR/proposal-data.json"
+
+echo "Turning the template into block proposal data using $BLOCK_TEMPLATE_TO_PROPOSAL..."
+
+# TODO: check the proposals for every valid time source: CurTime MinTime MaxTime ClampedNow
+cat "$TEMPLATE_NODE_TEMPLATE_RESPONSE" | \
+    $BLOCK_TEMPLATE_TO_PROPOSAL \
+        > "$PROPOSAL_DATA"
+
+# TODO: put this in a file if it exceeds the bash variable size limit
+PROPOSAL_ARG_FULL="{ \"mode\": \"proposal\", \"data\": \"$(cat $PROPOSAL_DATA)\" ${PROPOSAL_ARG:+, $PROPOSAL_ARG} }"
+
+echo "Block proposal data is in $PROPOSAL_DATA"
+# TODO: disable verbose block data
+cat "$PROPOSAL_DATA"
+
+echo
+echo
+
+echo "getblocktemplate proposal submissions:"
+echo "getblocktemplate $PROPOSAL_ARG_NO_DATA"
+echo
+
+PROPOSAL_NODES_PROPOSAL_RESPONSE_BASE="$ZCASH_RPC_TMP_DIR/proposal-check-getblocktemplate-proposal"
+PROPOSAL_NODES_PROPOSAL_RESPONSE_LIST=""
+
+for PORT in $PROPOSAL_RPC_PORTS; do
+    PROPOSAL_NODE_PROPOSAL_RESPONSE=$PROPOSAL_NODES_PROPOSAL_RESPONSE_BASE.$PORT.json
+    PROPOSAL_NODES_PROPOSAL_RESPONSE_LIST="$PROPOSAL_NODES_PROPOSAL_RESPONSE_LIST $PROPOSAL_NODE_PROPOSAL_RESPONSE"
+
+    # TODO: try modifying $PROPOSAL_DATA to contain the whole jsonrequestobject,
+    # and then call:
+    # xargs --max-chars=2000200 --arg-file="$PROPOSAL_DATA" time $ZCASH_CLI -rpcport="$PORT" getblocktemplate
+    time $ZCASH_CLI -rpcport="$PORT" getblocktemplate "$PROPOSAL_ARG_FULL" > "$PROPOSAL_NODE_PROPOSAL_RESPONSE"
+done
+
+echo
+
+if [ "$PROPOSAL_RPC_PORT_COUNT" -gt 1 ]; then
+    echo "Proposal response diffs between ports $PROPOSAL_RPC_PORTS:"
+
+    $DIFF $PROPOSAL_NODES_PROPOSAL_RESPONSE_LIST \
+        && ( \
+            echo "getblocktemplate proposal responses were identical"; \
+            echo ; \
+            echo "Final response from $PROPOSAL_NODE_PROPOSAL_RESPONSE:"; \
+            cat "$PROPOSAL_NODE_PROPOSAL_RESPONSE"; \
+            )
+else
+    echo "Proposal response from port $PROPOSAL_RPC_PORTS:"
+
+    cat $PROPOSAL_NODES_PROPOSAL_RESPONSE_LIST
```

Comment:
Optional: would be nice to have a message saying the block proposal was valid when the response is empty.

Reply:
Sounds good, I'll try to make this change some time today!

(For both the diff case and the single response case.)

Reply:
Added in commit 867e59f, rewritten in commit 947db58